### ISSUE TITLE
Multiple enumerations of IEnumerable<SqlDataRecord> passed as TVP

### ISSUE THREAD

**Issue by pvekshyn** [DapperLib/Dapper#2064]
Looks like Dapper twice enumerates IEnumerable<SqlDataRecord> passed as TVP
```
var numbers = GenerateNumbers(3);
BulkInsert(numbers);

static IEnumerable<int> GenerateNumbers(int count)
{
    for (int i = 0; i < count; i++)
    {
        Console.WriteLine(i);
        yield return i;
    }
}

void BulkInsert(IEnumerable<int> numbers)
{
    var records = numbers.Select(x => MapToSqlDataRecord(x));
    var sql = @"INSERT INTO [Test] (Id) SELECT Id FROM @tvp;";

    using (var connection = new SqlConnection(connectionString))
    {
        connection.Execute(sql, new { tvp = records.AsTableValuedParameter("dbo.TestType") });
    }
}

static SqlDataRecord MapToSqlDataRecord(int i)
{
    var record = new SqlDataRecord(
        new SqlMetaData("Id", SqlDbType.Int)
    );
    record.SetInt32(0, i);
    return record;
}
```
this code writes to console 
0 
0 
1 
2

and if I pass enumerable from database it throws exception because can not open second reader on same connection (same enumerable works fine using ADO.Net)

![image](https://github.com/DapperLib/Dapper/assets/119049273/13b67676-f288-4aac-be09-2334f43bdef2)
I believe this is the first enumeration

**Comment by pvekshyn** [comment]
Ah, I got the idea, you are trying to avoid exception when the enumerable is empty.
This is not correct, better catch exception in this case.

**Comment by mgravell** [comment]
I'm inclined to agree; at best we should type-test for `IReadOnlyCollection<T>` and perform this test via `.Count`; anything else: just send it down

**Comment by pvekshyn** [comment]
In this case it will not work for streaming enumerable to TVP, while it's perfectly working in ADO.Net

**Comment by pvekshyn** [comment]
Let's say I have any source of enumerable: db reader, http returning IAsyncEnumerable, grpc streaming, whatever
And I want to pass it to TVP without collecting the whole List in memory
I can do it using ADO.Net, but it's not possible using Dapper because of second enumeration, and it will be definitely not possible with IReadOnlyCollection

**Comment by mgravell** [comment]
yes, I understand and agree; the point is that what we should change here is basically:

```
if (source is IReadOnlyList<T> && source.Count == 0)
{
    // short-circuit
}
// main code
```

In your case, the type-test will fail, so we'll just fall into the `// main code`, without doing any iteration.

**Comment by VinceSeely** [comment]
@mgravell I would love to try giving this problem a stab and see if I can resolve the issue.

**Comment by bedamien** [comment]
Would using [`Enumerable.TryGetNonEnumeratedCount()`](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.trygetnonenumeratedcount) be worthwhile, over using a 'hand-rolled' check?
